packages/react/dialog/src/Dialog.tsx

```diff
@@ -247,31 +247,41 @@ const DialogContentModal = React.forwardRef((props, forwardedRef) => {
 
 const DialogContentNonModal = React.forwardRef((props, forwardedRef) => {
   const context = useDialogContext(CONTENT_NAME);
-  const isEscapeKeyDownRef = React.useRef(false);
+  const isInteractOutsideRef = React.useRef(false);
+
   return (
     <Portal>
       <DialogContentImpl
         {...props}
         ref={forwardedRef}
         trapFocus={false}
         disableOutsidePointerEvents={false}
-        onCloseAutoFocus={composeEventHandlers(props.onCloseAutoFocus, (event) => {
-          event.preventDefault();
-          if (isEscapeKeyDownRef.current) context.triggerRef.current?.focus();
-        })}
-        onEscapeKeyDown={composeEventHandlers(props.onEscapeKeyDown, () => {
-          isEscapeKeyDownRef.current = true;
-        })}
-        onInteractOutside={composeEventHandlers(props.onInteractOutside, (event) => {
+        onCloseAutoFocus={(event) => {
+          props.onCloseAutoFocus?.(event);
+          const userPrevented = event.defaultPrevented;
+
+          if (!userPrevented) {
+            event.preventDefault();
+            if (!isInteractOutsideRef.current) context.triggerRef.current?.focus();
+          }
+
+          isInteractOutsideRef.current = false;
+        }}
```

Comment:
Was there a reason we didn't go with this?

```suggestion
        onCloseAutoFocus={(event) => {
          props.onCloseAutoFocus?.(event);
          if (!event.defaultPrevented && !isInteractOutsideRef.current) {
			context.triggerRef.current?.focus();
          }

          isInteractOutsideRef.current = false;
          event.preventDefault();
        }}
```

The current one confused me because `preventDefault` is always happening either way. Either they are or we are so we may as well `event.preventDefault()` all the time. It didn't make sense why that had to be part of the conditional.

Reply:
In fact, shouldn't we only be preventing default if we're focusing ourselves? so it would be in **both** conditionals:

```jsx
        onCloseAutoFocus={(event) => {
          props.onCloseAutoFocus?.(event);
          if (!event.defaultPrevented && !isInteractOutsideRef.current) {
            event.preventDefault();
            context.triggerRef.current?.focus();
          }
          isInteractOutsideRef.current = false;
        }}
```

?

Reply:
Mainly for the ordering reason in Benoit's comment https://github.com/radix-ui/primitives/pull/799#discussion_r678099342 if we take it out of the condition then we need to ensure it comes after.

Reply:
> In fact, shouldn't we only be preventing default if we're focusing ourselves?

Ah I didn't see this comment before I added mine, yeah I think you're right!

Reply:
No, that's not right, we need to prevent **all the time** not just when we're focusing something manually. Like I explained here: https://github.com/radix-ui/primitives/pull/799#discussion_r678099342

> I understand we need to do it in ALL cases because either:
> - **we don't want to run any focus at all (ie. you've clicked somewhere outside, so natural focus should occur)**
> - we want explicit focus to trigger so we prevent the default close auto focus

Reply:
@benoitgrelard is on the money here, it always needs to happen in non-modal to avoid focus fighting on outside clicks etc. https://github.com/radix-ui/primitives/pull/799/commits/fb6e32cca44c9a271a9ada0db4d75713e230e308 this whole thing got a little confusing after a while 😅

Reply:
Ah of course... wow, yeah, my head is spinning 😅

Reply:
On the whole "coming after" thing. What's wrong with this? In the olden days this was actually the recommended place to put it to ensure progressive enhancement didn't break if there was a JS error in the handling logic.

Reply:
See this example where when JS fails, the app still works. **if the event is prevented after** https://codepen.io/pen/?editors=1010